test_mapbased_param/Cl_estimation.py

```diff
@@ -0,0 +1,375 @@
+                                no_inh=self.config['no_inh'])
+
+        # reformating the simulated noise maps 
+        noise_maps_ = np.zeros((n_cov.shape[0], 3, W.shape[-1]))
+        ind = 0
+        for f in range(n_cov.shape[0]): 
+            for i in range(3): 
+                noise_maps_[f,i,:] += noise_maps_sim[ind,:]*1.0
+                ind += 1
+        # only keeping Q and U
+        noise_maps_ = noise_maps_[:,1:,:]
+        # propagate noise through the map-making equation
+        Q_noise_cmb = np.einsum('fp,fp->p', W[0,:,0,:], noise_maps_[:,0])
+        U_noise_cmb = np.einsum('fp,fp->p', W[0,:,1,:], noise_maps_[:,1])
+        # compute corresponding spectra
+        fn = get_field_func(mask*Q_noise_cmb, mask*U_noise_cmb, mask_apo)
+        Cl_noise_bias.append(Cl_func(fn, fn, w)[3] )
+
+    return Cl_noise_bias
+
+
+class BBClEstimation(PipelineStage):
+    """
+    Stage that performs estimate the angular power spectra of:
+        * each of the sky components, foregrounds and CMB, as well as the cross terms
+        * of the corresponding covariance
+    """
+
+    name='BBClEstimation'
+    inputs=[('binary_mask_cut',FitsFile),('post_compsep_maps',FitsFile), ('post_compsep_cov',FitsFile),
+            ('A_maxL',NumpyFile),('noise_maps',FitsFile), ('post_compsep_noise',FitsFile), 
+            ('norm_hits_map', FitsFile), ('frequency_maps',FitsFile),('CMB_template_150GHz', FitsFile),\
+            ('mask_patches', FitsFile),('noise_cov',FitsFile)]
+    
+    outputs=[('Cl_clean', FitsFile),('Cl_noise', FitsFile),('Cl_cov_clean', FitsFile), 
+             ('Cl_cov_freq', FitsFile), ('fsky_eff',TextFile), ('Cl_fgs', NumpyFile),
+             ('Cl_CMB_template_150GHz', NumpyFile), ('mask_apo', FitsFile),
+             ('Cl_noise_bias', FitsFile)]
+
+    def run(self):
+
+        clean_map = hp.read_map(self.get_input('post_compsep_maps'),verbose=False, field=None, h=False)
+        cov_map = hp.read_map(self.get_input('post_compsep_cov'),verbose=False, field=None, h=False)
+        A_maxL = np.load(self.get_input('A_maxL'))
+        noise_maps=hp.read_map(self.get_input('noise_maps'),verbose=False, field=None)
+        post_compsep_noise=hp.read_map(self.get_input('post_compsep_noise'),verbose=False, field=None)
+        frequency_maps=hp.read_map(self.get_input('frequency_maps'),verbose=False, field=None)
+        CMB_template_150GHz = hp.read_map(self.get_input('CMB_template_150GHz'), field=None)
+        
+        nhits_raw = hp.read_map(self.get_input('norm_hits_map'))
+        nhits = hp.ud_grade(nhits_raw,nside_out=self.config['nside'])
+        nh = mknm.get_mask(nhits, nside_out=self.config['nside'])
+
+        mask_patches = hp.read_map(self.get_input('mask_patches'), verbose=False, field=None)
+        noise_cov=hp.read_map(self.get_input('noise_cov'),verbose=False, field=None)
+        # reorganization of noise covariance 
+        instrument = {'frequencies':np.array(self.config['frequencies'])}
+        ind = 0
```

Comment:
I think you should be able to do this as: 
```
noise_cov = noise_cov.reshape([len(instrument['frequencies']),
                                                3, frequency_maps.shape[-1]])
```
That way you also wouldn't have to store both `noise_cov` and `noise_cov_` in memory.